File: src/security/network_policy_manager.rs
use anyhow::Result;
use std::collections::HashMap;
use std::net::IpAddr;
use std::sync::Arc;
use tokio::sync::Mutex;
use log::{info, warn, error, debug, trace};
use std::time::{SystemTime, UNIX_EPOCH};

use crate::network::{NetworkPolicy, NetworkRule, NetworkSelector, Protocol, PolicyAction};
use crate::security::network_policy_controller::{NetworkPolicyController, ViolationSeverity};
use crate::monitoring::MonitoringManager;

/// NetworkPolicyManager handles network policies using the NetworkPolicyController
#[derive(Debug)]
pub struct NetworkPolicyManager {
    controller: Arc<Mutex<NetworkPolicyController>>,
    monitoring_manager: Option<Arc<MonitoringManager>>,
    notification_endpoints: Vec<String>,
    dynamic_policy_adjustment: bool,
    violation_threshold: u32,
    adjustment_cooldown_seconds: u64,
    last_adjustment: Arc<Mutex<HashMap<String, u64>>>,
}

impl NetworkPolicyManager {
    pub fn new() -> Self {
        Self {
            controller: Arc::new(Mutex::new(NetworkPolicyController::new())),
            monitoring_manager: None,
            notification_endpoints: Vec::new(),
            dynamic_policy_adjustment: false,
            violation_threshold: 5, // Default: adjust policy after 5 violations
            adjustment_cooldown_seconds: 3600, // Default: 1 hour cooldown between adjustments
            last_adjustment: Arc::new(Mutex::new(HashMap::new())),
        }
    }
    
    /// Set monitoring manager for metrics collection
    pub fn with_monitoring_manager(mut self, monitoring_manager: Arc<MonitoringManager>) -> Self {
        // Set monitoring manager for the policy manager
        self.monitoring_manager = Some(monitoring_manager.clone());
        
        // Also set it for the controller
        let controller = self.controller.clone();
        tokio::spawn(async move {
            let mut controller_lock = controller.lock().await;
            *controller_lock = controller_lock.clone().with_monitoring_manager(monitoring_manager);
        });
        
        self
    }
    
    /// Add notification endpoint for policy violation alerts
    pub async fn add_notification_endpoint(&mut self, endpoint: String) -> Result<()> {
        // Add endpoint to the policy manager
        self.notification_endpoints.push(endpoint.clone());
        
        // Also add it to the controller
        let mut controller = self.controller.lock().await;
        controller.add_notification_endpoint(endpoint);
        
        Ok(())
    }
    
    /// Enable or disable dynamic policy adjustment
    pub fn set_dynamic_policy_adjustment(&mut self, enabled: bool) {
        self.dynamic_policy_adjustment = enabled;
    }
    
    /// Set the violation threshold for dynamic policy adjustment
    pub fn set_violation_threshold(&mut self, threshold: u32) {
        self.violation_threshold = threshold;
    }
    
    /// Set the cooldown period between policy adjustments
    pub fn set_adjustment_cooldown(&mut self, seconds: u64) {
        self.adjustment_cooldown_seconds = seconds;
    }
    
    /// Register container labels for policy matching
    pub async fn register_container_labels(&mut self, container_id: &str, labels: HashMap<String, String>) -> Result<()> {
        // We'll store the labels but wait for the IP to be registered before applying policies
        let mut controller = self.controller.lock().await;
        
        // Check if we already have an IP for this container
        if let Some(ip) = controller.get_container_ip(container_id).await {
            // If we have both IP and labels, register the container
            controller.register_container(container_id, ip, labels).await?;
        } else {
            // Otherwise just store the labels
            controller.store_container_labels(container_id, labels).await?;
        }
        
        Ok(())
    }
    
    /// Register container IP address
    pub async fn register_container_ip(&mut self, container_id: &str, ip: IpAddr) -> Result<()> {
        let mut controller = self.controller.lock().await;
        
        // Check if we already have labels for this container
        if let Some(labels) = controller.get_container_labels(container_id).await {
            // If we have both IP and labels, register the container
            controller.register_container(container_id, ip, labels).await?;
        } else {
            // Otherwise just store the IP
            controller.store_container_ip(container_id, ip).await?;
        }
        
        Ok(())
    }
    
    /// Unregister container when it's removed
    pub async fn unregister_container(&mut self, container_id: &str) -> Result<()> {
        self.controller.lock().await.unregister_container(container_id).await?;
        Ok(())
    }
    
    /// Apply a network policy
    pub async fn apply_policy(&mut self, policy: NetworkPolicy) -> Result<()> {
        debug!("Applying network policy: {}", policy.name);
        
        // Apply the policy using the controller
        self.controller.lock().await.apply_policy(policy.clone()).await?;
        
        // Send metrics if monitoring is available
        if let Some(monitoring) = &self.monitoring_manager {
            let mut labels = HashMap::new();
            labels.insert("policy_name".to_string(), policy.name.clone());
            if let Some(tenant_id) = &policy.tenant_id {
                labels.insert("tenant_id".to_string(), tenant_id.clone());
            }
            
            let metrics = vec![
                crate::monitoring::Metric {
                    name: "network_policy_applied".to_string(),
                    value: 1.0,
                    labels,
                    timestamp: SystemTime::now()
                        .duration_since(UNIX_EPOCH)
                        .unwrap_or_default()
                        .as_secs() as i64,
                },
            ];
            
            if let Err(e) = monitoring.send_metrics(metrics).await {
                warn!("Failed to send network policy metrics: {}", e);
            }
            
            /// Dynamically adjust a network policy based on violation patterns
            pub async fn adjust_policy_for_violations(&mut self, policy_name: &str) -> Result<bool> {
                // Skip if dynamic adjustment is disabled
                if !self.dynamic_policy_adjustment {
                    return Ok(false);
                }
                
                // Get current time
                let now = SystemTime::now()
                    .duration_since(UNIX_EPOCH)
                    .unwrap_or_default()
                    .as_secs();
                    
                // Check cooldown period
                let mut last_adjustment = self.last_adjustment.lock().await;
                if let Some(last_time) = last_adjustment.get(policy_name) {
                    if now - *last_time < self.adjustment_cooldown_seconds {
                        debug!("Skipping policy adjustment for {} due to cooldown period", policy_name);
                        return Ok(false);
                    }
                }
                
                // Get violation logs for this policy
                let controller = self.controller.lock().await;
                let violations = controller.get_violation_logs().await?;
                
                // Count recent violations for this policy
                let recent_violations = violations.iter()
                    .filter(|v| v.policy_name == policy_name)
                    .filter(|v| (now - v.timestamp) < 24 * 60 * 60) // Last 24 hours
                    .count();
                    
                // If violations exceed threshold, adjust the policy
                if recent_violations >= self.violation_threshold as usize {
                    info!("Detected {} violations for policy {}, adjusting policy",
                        recent_violations, policy_name);
                        
                    // Get the current policy
                    let policies = controller.get_policies().await?;
                    if let Some(mut policy) = policies.get(policy_name).cloned() {
                        // Adjust the policy based on violation patterns
                        self.strengthen_policy(&mut policy, &violations).await?;
                        
                        // Apply the updated policy
                        drop(controller); // Release the lock before calling apply_policy
                        self.apply_policy(policy.clone()).await?;
                        
                        // Update last adjustment time
                        last_adjustment.insert(policy_name.to_string(), now);
                        
                        // Log the adjustment
                        info!("Policy {} adjusted due to {} recent violations", policy_name, recent_violations);
                        
                        // Send metrics if monitoring is available
                        if let Some(monitoring) = &self.monitoring_manager {
                            let mut labels = HashMap::new();
                            labels.insert("policy_name".to_string(), policy_name.to_string());
                            labels.insert("violation_count".to_string(), recent_violations.to_string());
                            
                            let metrics = vec![
                                crate::monitoring::Metric {
                                    name: "network_policy_adjusted".to_string(),
                                    value: 1.0,
                                    labels,
                                    timestamp: now as i64,
                                },
                            ];
                            
                            if let Err(e) = monitoring.send_metrics(metrics).await {
                                warn!("Failed to send policy adjustment metrics: {}", e);
                            }
                        }
                        
                        return Ok(true);
                    }
                }
                
                Ok(false)
            }
            
            /// Strengthen a policy based on violation patterns
            async fn strengthen_policy(&self, policy: &mut NetworkPolicy, violations: &[crate::security::network_policy_controller::ViolationLog]) -> Result<()> {
                // Group violations by source/destination patterns
                let mut violation_patterns: HashMap<(String, String), Vec<&crate::security::network_policy_controller::ViolationLog>> = HashMap::new();
                
                for violation in violations.iter().filter(|v| v.policy_name == policy.name) {
                    let src_pattern = violation.source_container.clone();
                    let dst_pattern = violation.destination_container.clone().unwrap_or_else(|| "unknown".to_string());
                    
                    violation_patterns
                        .entry((src_pattern, dst_pattern))
                        .or_insert_with(Vec::new)
                        .push(violation);
                }
                
                // Find the most common violation patterns
                let mut pattern_counts: Vec<((String, String), usize)> = violation_patterns
                    .iter()
                    .map(|(pattern, violations)| (pattern.clone(), violations.len()))
                    .collect();
                    
                pattern_counts.sort_by(|a, b| b.1.cmp(&a.1));
                
                // Add or strengthen rules for the top violation patterns
                for ((src_pattern, dst_pattern), count) in pattern_counts.iter().take(3) {
                    if *count >= 3 {
                        debug!("Adding stricter rule for pattern: {} -> {} ({} violations)",
                            src_pattern, dst_pattern, count);
                        
                        // Create a new rule or strengthen existing rules
                        let mut rule_added = false;
                        
                        for rule in &mut policy.ingress_rules {
                            // If we find a matching rule, make it stricter
                            if rule.from.iter().any(|peer| {
                                if let Some(selector) = &peer.selector {
                                    selector.labels.values().any(|v| v.contains(src_pattern))
                                } else {
                                    false
                                }
                            }) {
                                // Make the rule stricter
                                rule.action = Some(PolicyAction::Deny);
                                rule.log = true;
                                rule_added = true;
                                break;
                            }
                        }
                        
                        // If no matching rule found, add a new one
                        if !rule_added {
                            // Create a new selector for the source pattern
                            let mut selector_labels = HashMap::new();
                            
                            // Try to extract meaningful labels from the pattern
                            if src_pattern.contains("frontend") {
                                selector_labels.insert("app".to_string(), "frontend".to_string());
                            } else if src_pattern.contains("backend") {
                                selector_labels.insert("app".to_string(), "backend".to_string());
                            } else {
                                // Use a generic label based on the pattern
                                selector_labels.insert("name".to_string(), src_pattern.clone());
                            }
                            
                            // Create a new rule
                            let new_rule = NetworkRule {
                                ports: Vec::new(), // Empty means all ports
                                from: vec![
                                    crate::network::NetworkPeer {
                                        ip_block: None,
                                        selector: Some(crate::network::NetworkSelector {
                                            labels: selector_labels,
                                        }),
                                    },
                                ],
                                action: Some(PolicyAction::Deny),
                                log: true,
                                description: Some(format!("Auto-generated rule based on {} violation patterns", count)),
                                id: Some(format!("auto-{}", uuid::Uuid::new_v4())),
                            };
                            
                            // Add the new rule
                            policy.ingress_rules.push(new_rule);
                        }
                    }
                }
                
                // Increase policy priority to ensure it takes precedence
                policy.priority += 10;
                
                Ok(())
            }
            
            /// Run maintenance tasks for network policies
            pub async fn run_maintenance(&mut self) -> Result<()> {
                debug!("Running network policy maintenance tasks");
                
                // Run reconciliation to ensure policies are consistently applied
                self.run_reconciliation().await?;
                
                // Check for policy violations
                let controller = self.controller.lock().await;
                let violations = controller.get_violation_logs().await?;
                drop(controller);
                
                // Count violations by policy
                let mut policy_violations: HashMap<String, usize> = HashMap::new();
                
                for violation in &violations {
                    *policy_violations.entry(violation.policy_name.clone()).or_insert(0) += 1;
                }
                
                // Adjust policies with high violation counts
                for (policy_name, count) in policy_violations {
                    if count >= self.violation_threshold as usize {
                        self.adjust_policy_for_violations(&policy_name).await?;
                    }
                }
                
                // Clean up old violation logs (older than 30 days)
                let controller = self.controller.lock().await;
                let removed = controller.cleanup_old_violations(30).await?;
                if removed > 0 {
                    debug!("Cleaned up {} old violation logs during maintenance", removed);
                }
                
                info!("Network policy maintenance completed");
                
                Ok(())
            }
        }
        
        Ok(())
    }
    
    /// Run reconciliation loop to ensure policies are consistently applied
    pub async fn run_reconciliation(&mut self) -> Result<()> {
        self.controller.lock().await.run_reconciliation().await?;
        Ok(())
    }
    
    /// Create default network policies for a new tenant
    pub async fn create_default_tenant_policies(&mut self, tenant_id: &str) -> Result<()> {
        self.controller.lock().await.create_default_tenant_policies(tenant_id).await?;
        Ok(())
    }
    
    /// Check if traffic is allowed between containers
    pub async fn is_traffic_allowed(&self, src_container_id: &str, dst_container_id: &str, protocol: Protocol, port: u16) -> Result<bool> {
        Ok(self.controller.lock().await.is_traffic_allowed(src_container_id, dst_container_id, protocol, port).await?)
    }
    
    /// Get policy violation logs
    pub async fn get_violation_logs(&self) -> Result<Vec<crate::security::network_policy_controller::ViolationLog>> {
        Ok(self.controller.lock().await.get_violation_logs().await)
    }
}

File: internal/security/network_policy_manager.go
package security

import (
	"context"
	"fmt"
	"net"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/sirupsen/logrus"
)

// NetworkPolicyManager handles network policies using the NetworkPolicyController
type NetworkPolicyManager struct {
	controller              *NetworkPolicyController
	mutex                   sync.RWMutex
	dynamicPolicyAdjustment bool
	violationThreshold      uint32
	adjustmentCooldown      time.Duration
	lastAdjustment          map[string]time.Time
	logger                  *logrus.Logger
}

// NewNetworkPolicyManager creates a new network policy manager
func NewNetworkPolicyManager(logger *logrus.Logger) *NetworkPolicyManager {
	if logger == nil {
		logger = logrus.New()
		logger.SetLevel(logrus.InfoLevel)
	}

	return &NetworkPolicyManager{
		controller:              NewNetworkPolicyController(logger),
		dynamicPolicyAdjustment: false,
		violationThreshold:      5,         // Default: adjust policy after 5 violations
		adjustmentCooldown:      time.Hour, // Default: 1 hour cooldown between adjustments
		lastAdjustment:          make(map[string]time.Time),
		logger:                  logger,
	}
}

// WithNetworkPolicyController sets the network policy controller for the manager
func (m *NetworkPolicyManager) WithNetworkPolicyController(controller *NetworkPolicyController) {
	m.mutex.Lock()
	defer m.mutex.Unlock()

	m.controller = controller
	m.logger.Debug("Network policy controller has been set")
}

// AddNotificationEndpoint adds a notification endpoint for policy violation alerts
func (m *NetworkPolicyManager) AddNotificationEndpoint(ctx context.Context, endpoint string) error {
	m.mutex.Lock()
	defer m.mutex.Unlock()

	m.controller.AddNotificationEndpoint(endpoint)
	m.logger.WithField("endpoint", endpoint).Info("Added notification endpoint")
	return nil
}

// SetDynamicPolicyAdjustment enables or disables dynamic policy adjustment
func (m *NetworkPolicyManager) SetDynamicPolicyAdjustment(enabled bool) {
	m.mutex.Lock()
	defer m.mutex.Unlock()

	m.dynamicPolicyAdjustment = enabled
	m.logger.WithField("enabled", enabled).Info("Set dynamic policy adjustment")
}

// SetViolationThreshold sets the violation threshold for dynamic policy adjustment
func (m *NetworkPolicyManager) SetViolationThreshold(threshold uint32) {
	m.mutex.Lock()
	defer m.mutex.Unlock()

	m.violationThreshold = threshold
	m.logger.WithField("threshold", threshold).Info("Set violation threshold")
}

// SetAdjustmentCooldown sets the cooldown period between policy adjustments
func (m *NetworkPolicyManager) SetAdjustmentCooldown(cooldown time.Duration) {
	m.mutex.Lock()
	defer m.mutex.Unlock()

	m.adjustmentCooldown = cooldown
	m.logger.WithField("cooldown", cooldown).Info("Set adjustment cooldown")
}

// RegisterContainerLabels registers container labels for policy matching
func (m *NetworkPolicyManager) RegisterContainerLabels(ctx context.Context, containerID string, labels map[string]string) error {
	m.mutex.Lock()
	defer m.mutex.Unlock()

	// Check if we already have an IP for this container
	ip, exists := m.controller.GetContainerIP(containerID)
	if exists {
		// If we have both IP and labels, register the container
		return m.controller.RegisterContainer(ctx, containerID, ip, labels)
	}

	// Otherwise just store the labels
	return m.controller.StoreContainerLabels(ctx, containerID, labels)
}

// RegisterContainerIP registers a container IP address
func (m *NetworkPolicyManager) RegisterContainerIP(ctx context.Context, containerID string, ip net.IP) error {
	m.mutex.Lock()
	defer m.mutex.Unlock()

	// Check if we already have labels for this container
	labels, exists := m.controller.GetContainerLabels(containerID)
	if exists {
		// If we have both IP and labels, register the container
		return m.controller.RegisterContainer(ctx, containerID, ip, labels)
	}

	// Otherwise just store the IP
	return m.controller.StoreContainerIP(ctx, containerID, ip)
}

// UnregisterContainer unregisters a container when it's removed
func (m *NetworkPolicyManager) UnregisterContainer(ctx context.Context, containerID string) error {
	m.mutex.Lock()
	defer m.mutex.Unlock()

	return m.controller.UnregisterContainer(ctx, containerID)
}

// ApplyPolicy applies a network policy
func (m *NetworkPolicyManager) ApplyPolicy(ctx context.Context, policy NetworkPolicy) error {
	m.mutex.Lock()
	defer m.mutex.Unlock()

	m.logger.WithField("policy_name", policy.Name).Info("Applying network policy")

	// Apply the policy using the controller
	if err := m.controller.ApplyPolicy(ctx, policy); err != nil {
		return fmt.Errorf("failed to apply policy: %w", err)
	}

	return nil
}

// AdjustPolicyForViolations dynamically adjusts a network policy based on violation patterns
func (m *NetworkPolicyManager) AdjustPolicyForViolations(ctx context.Context, policyName string) (bool, error) {
	m.mutex.Lock()
	defer m.mutex.Unlock()

	// Skip if dynamic adjustment is disabled
	if !m.dynamicPolicyAdjustment {
		return false, nil
	}

	// Check cooldown period
	now := time.Now()
	lastTime, exists := m.lastAdjustment[policyName]
	if exists && now.Sub(lastTime) < m.adjustmentCooldown {
		m.logger.WithField("policy_name", policyName).Debug("Skipping policy adjustment due to cooldown period")
		return false, nil
	}

	// Get violation logs
	violations := m.controller.GetViolationLogs(ctx)

	// Count recent violations for this policy
	recentViolations := 0
	for _, v := range violations {
		if v.PolicyName == policyName && time.Since(time.Unix(v.Timestamp, 0)) < 24*time.Hour {
			recentViolations++
		}
	}

	// If violations exceed threshold, adjust the policy
	if uint32(recentViolations) >= m.violationThreshold {
		m.logger.WithFields(logrus.Fields{
			"policy_name":     policyName,
			"violation_count": recentViolations,
		}).Info("Detected violations exceeding threshold, adjusting policy")

		// Get the current policy
		policy, err := m.controller.GetPolicy(ctx, policyName)
		if err != nil {
			return false, fmt.Errorf("failed to get policy: %w", err)
		}

		// Adjust the policy based on violation patterns
		if err := m.strengthenPolicy(ctx, policy, violations); err != nil {
			return false, fmt.Errorf("failed to strengthen policy: %w", err)
		}

		// Apply the updated policy
		if err := m.controller.ApplyPolicy(ctx, *policy); err != nil {
			return false, fmt.Errorf("failed to apply adjusted policy: %w", err)
		}

		// Update last adjustment time
		m.lastAdjustment[policyName] = now

		m.logger.WithFields(logrus.Fields{
			"policy_name":     policyName,
			"violation_count": recentViolations,
		}).Info("Policy adjusted due to violations")

		return true, nil
	}

	return false, nil
}

// strengthenPolicy strengthens a policy based on violation patterns
func (m *NetworkPolicyManager) strengthenPolicy(ctx context.Context, policy *NetworkPolicy, violations []ViolationLog) error {
	// Group violations by source/destination patterns
	violationPatterns := make(map[string][]ViolationLog)

	for _, violation := range violations {
		if violation.PolicyName == policy.Name {
			srcPattern := violation.SourceContainer
			dstPattern := "unknown"
			if violation.DestinationContainer != nil {
				dstPattern = *violation.DestinationContainer
			}
			pattern := fmt.Sprintf("%s->%s", srcPattern, dstPattern)

			violationPatterns[pattern] = append(violationPatterns[pattern], violation)
		}
	}

	// Find the most common violation patterns
	type patternCount struct {
		pattern string
		count   int
	}
	patternCounts := make([]patternCount, 0, len(violationPatterns))
	for pattern, violations := range violationPatterns {
		patternCounts = append(patternCounts, patternCount{
			pattern: pattern,
			count:   len(violations),
		})
	}

	// Sort by count in descending order
	sort.Slice(patternCounts, func(i, j int) bool {
		return patternCounts[i].count > patternCounts[j].count
	})

	// Add or strengthen rules for the top violation patterns
	for i, pc := range patternCounts {
		if i >= 3 || pc.count < 3 {
			break
		}

		pattern := pc.pattern
		count := pc.count

		m.logger.WithFields(logrus.Fields{
			"pattern": pattern,
			"count":   count,
		}).Debug("Adding stricter rule for pattern")

		// Extract source pattern from the combined pattern
		parts := strings.Split(pattern, "->")
		srcPattern := parts[0]

		// Create a new rule or strengthen existing rules
		ruleAdded := false

		for i := range policy.IngressRules {
			rule := &policy.IngressRules[i]
			// If we find a matching rule, make it stricter
			for _, peer := range rule.From {
				if peer.Selector != nil {
					for _, v := range peer.Selector.Labels {
						if strings.Contains(v, srcPattern) {
							// Make the rule stricter
							denyAction := PolicyActionDeny
							rule.Action = &denyAction
							rule.Log = true
							ruleAdded = true
							break
						}
					}
				}
				if ruleAdded {
					break
				}
			}
			if ruleAdded {
				break
			}
		}

		// If no matching rule found, add a new one
		if !ruleAdded {
			// Create a new selector for the source pattern
			selectorLabels := make(map[string]string)

			// Try to extract meaningful labels from the pattern
			if strings.Contains(srcPattern, "frontend") {
				selectorLabels["app"] = "frontend"
			} else if strings.Contains(srcPattern, "backend") {
				selectorLabels["app"] = "backend"
			} else {
				// Use a generic label based on the pattern
				selectorLabels["name"] = srcPattern
			}

			// Create a new rule
			denyAction := PolicyActionDeny
			newRule := NetworkRule{
				Ports: []PortRange{}, // Empty means all ports
				From: []NetworkPeer{
					{
						Selector: &NetworkSelector{
							Labels: selectorLabels,
						},
					},
				},
				Action:      &denyAction,
				Log:         true,
				Description: fmt.Sprintf("Auto-generated rule based on %d violation patterns", count),
				ID:          fmt.Sprintf("auto-%s", uuid.New().String()),
			}

			// Add the new rule
			policy.IngressRules = append(policy.IngressRules, newRule)
		}
	}

	// Increase policy priority to ensure it takes precedence
	policy.Priority += 10

	return nil
}

// RunMaintenance runs maintenance tasks for network policies
func (m *NetworkPolicyManager) RunMaintenance(ctx context.Context) error {
	m.logger.Info("Running network policy maintenance tasks")

	// Run reconciliation to ensure policies are consistently applied
	if err := m.RunReconciliation(ctx); err != nil {
		return fmt.Errorf("failed to run reconciliation: %w", err)
	}

	// Get violation logs
	violations := m.controller.GetSecurityEvents(ctx)

	// Count violations by policy
	policyViolations := make(map[string]int)
	for _, violation := range violations {
		if violation.EventType == EventTypePolicyViolation {
			policyName, exists := violation.Details["policy_name"]
			if exists {
				policyViolations[policyName]++
			}
		}
	}

	// Adjust policies with high violation counts
	for policyName, count := range policyViolations {
		if uint32(count) >= m.violationThreshold {
			if _, err := m.AdjustPolicyForViolations(ctx, policyName); err != nil {
				m.logger.WithError(err).WithField("policy_name", policyName).Error("Failed to adjust policy")
			}
		}
	}

	// Clean up old violation logs (older than 30 days)
	removed, err := m.controller.CleanupOldViolations(ctx, 30)
	if err != nil {
		m.logger.WithError(err).Error("Failed to clean up old violation logs")
	} else if removed > 0 {
		m.logger.WithField("removed_count", removed).Debug("Cleaned up old violation logs")
	}

	m.logger.Info("Network policy maintenance completed")

	return nil
}

// RunReconciliation runs the reconciliation loop to ensure policies are consistently applied
func (m *NetworkPolicyManager) RunReconciliation(ctx context.Context) error {
	return m.controller.RunReconciliation(ctx)
}

// CreateDefaultTenantPolicies creates default network policies for a new tenant
func (m *NetworkPolicyManager) CreateDefaultTenantPolicies(ctx context.Context, tenantID string) error {
	return m.controller.CreateDefaultTenantPolicies(ctx, tenantID)
}

// IsTrafficAllowed checks if traffic is allowed between containers
func (m *NetworkPolicyManager) IsTrafficAllowed(ctx context.Context, srcContainerID, dstContainerID string, protocol Protocol, port uint16) (bool, error) {
	return m.controller.IsTrafficAllowed(ctx, srcContainerID, dstContainerID, protocol, port)
}

// GetViolationLogs gets policy violation logs
func (m *NetworkPolicyManager) GetViolationLogs(ctx context.Context, tenantID string) ([]ViolationLog, error) {
	return m.controller.GetTenantViolationLogs(ctx, tenantID)
}
